client/allocrunner/taskrunner/task_runner.go

```diff
@@ -705,13 +705,17 @@ func (tr *TaskRunner) runDriver() error {
 		tr.logger.Warn("some environment variables not available for rendering", "keys", strings.Join(keys, ", "))
 	}
 
-	val, diag := hclutils.ParseHclInterface(tr.task.Config, tr.taskSchema, vars)
+	val, diag, diagErrs := hclutils.ParseHclInterface(tr.task.Config, tr.taskSchema, vars)
 	if diag.HasErrors() {
-		return multierror.Append(errors.New("failed to parse config"), diag.Errs()...)
+		parseErr := multierror.Append(errors.New("failed to parse config"), diagErrs...)
+		tr.EmitEvent(structs.NewTaskEvent(structs.TaskFailedValidation).SetDriverError(parseErr))
```

Comment:
I see that in the snippet that this returns a generic `Validation of task failed` as the message.  It's because `TaskFailedValidation` expects `ValidationError`[1] to be set instead of `DriverError`.

Also, using the `multierror.Append` generates very awkward messages, like:

```
2019-07-18T10:22:30+07:00  Failed Validation  2 error(s) occurred:

* failed to parse config
* Invalid label: No argument or block type is named "bad_key".
```

It'd be nice to format the errors ourselves and present something like the following

```
2019-07-18T10:22:30+07:00  Failed Validation  failed to parse config; found the following:

* Invalid label: No argument or block type is named "bad_key".
```

[1] https://github.com/hashicorp/nomad/blob/87d716ed0b363d0e32dfa71eae1d8d2714a4455d/command/alloc_status.go#L374-L379

Reply:
Going to address better multierror formatting in another PR